kicad: add srcs parameter to allow configuring kicad versions

matthuszagh opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Allows overriding kicad versions which are otherwise inaccessible.

###### Things done
Added srcs parameter that can be used to customize kicad or kicad-unstable versions, including libraries. For instance, you could do something like:
```nix
final: prev:

{
  kicad-unstable = (prev.kicad-unstable.override {
    srcs = {
      "kicad-unstable" = {
        kicadVersion = {
          version = "2020-10-08";
          src = {
            rev = "fd22fe8e374ce71d57e9f683ba996651aa69fa4e";
            sha256 = "sha256-F8qugru/jU3DgZSpQXQhRGNFSk0ybFRkpyWb7HAGBdc=";
          };
        };
        libVersion = {
          version = "2020-08-22";
          libSources = {
            i18n.rev = "cbbb1efd940094bf0c3168280698b2b059a8c509";
            i18n.sha256 = "1q4jakn6m8smnr2mg7jgb520nrb6fag9mdvlcpx3smp3qbxka818";
            symbols.rev = "9ca6a5348cdeb88e699582d4ed051ff7303b44d3";
            symbols.sha256 = "13w6pb34rhz96rnar25z7kiscy6q1fm8l39hq1bpb8g9yn86ssz4";
            templates.rev = "ae16953b81055855bcede4a33305413599d86a15";
            templates.sha256 = "1pkv90p3liy3bj4nklxsvpzh9m56p0k5ldr22armvgqfaqaadx9v";
            footprints.rev = "f94c2d5d619d16033f69a555b449f59604d97865";
            footprints.sha256 = "1g71sk77jvqaf9xvgq6dkyvd9pij2lb4n0bn0dqnwddhwam935db";
            packages3d.rev = "f699b0e3c13fe75618086913e39279c85da14cc7";
            packages3d.sha256 = "0m5rb5axa946v729z35ga84in76y4zpk32qzi0hwqx957zy72hs9";
          };
        };
      };
    };
}
```

I've also reformatted the existing code. Let me know if the reformatting is too invasive and I'll revert.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

doronbehar commented:
@matthuszagh this change is somewhat simple enough so it's easy for me to press the button, but I think it would be nice if it was possible to use a different _fork_ of a library / kicad itself, not only a different revision.

P.S the eval errors have nothing to do with you.

matthuszagh commented:
I didn't consider that but I agree with you. Let me dig into this a little more.

matthuszagh commented:
@doronbehar this now allows overriding with a full `fetchFromGitLab`, or whatever you want. Versions can also be overriden with `kicadVersion` or `libVersion`. For example, this is how I'm using it:

```nix
final: prev:

{
  kicad-unstable = (prev.kicad-unstable.override {
    symbol-path = "/home/matt/src/kicad-symbols";
    footprint-path = "/home/matt/src/kicad-footprints";
    model3d-path = "/home/matt/src/kicad-packages3D";
    srcs = {
      kicadVersion = "2020-10-08";
      kicad = prev.fetchFromGitLab {
        group = "kicad";
        owner = "code";
        repo = "kicad";
        rev = "fd22fe8e374ce71d57e9f683ba996651aa69fa4e";
        sha256 = "sha256-F8qugru/jU3DgZSpQXQhRGNFSk0ybFRkpyWb7HAGBdc=";
      };
    };
  });
}
```

Let me know what you think.

doronbehar commented on pkgs/applications/science/electronics/kicad/libraries.nix:
```diff
@@ -1,39 +1,27 @@
-#   };
-# };
-with lib;
+with stdenv.lib;
```
https://nix.dev/anti-patterns/language.html#with-attrset-expression

doronbehar commented on pkgs/applications/science/electronics/kicad/i18n.nix:
```diff
@@ -0,0 +1,19 @@
+{ stdenv
+, cmake
+, gettext
+, i18nSrc
+, i18nVersion
```
Nitpick: Use `src` and `version` here, and `inherit src version` under mkDerivation.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,52 +1,119 @@
+    sha256 = versionsImport.${baseName}.libVersion.libSources.${name}.sha256;
+  };
+
+  srcOverridep = attr: (!stable && srcs ? ${attr});
```
More readble IMO:

```suggestion
  srcOverridep = attr: (!stable && builtins.hasAttr attr srcs);
```

doronbehar reviewed (commented):
Besides these, here are general comments:

1. It'd be easier to review this if the indentation settings will be in a separate commit.
2. More comments to how the logic is going on inside `default.nix` regarding `srcs`.
3. Since there's no way to query a package's override options, it'd be nice to put a big comment at the start of `default.nix` with all the possible ways one can override `srcs`.

matthuszagh committed:
kicad: reformat with nixpkgs-fmt

doronbehar commented:
Oh and 1 more thing: I'm rather sure this change shouldn't trigger rebuilds - ofborg detects it [this way](https://github.com/NixOS/ofborg#running-meta-checks-locally) and in the "eval" check of the PR, there's a link to the list of changed outputs. Also the PR's labels ofborg sets reflects that.

matthuszagh commented:
@doronbehar thanks for the review. I've applied the changes. Let me know.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,52 +1,165 @@
 stdenv.mkDerivation rec {
 
-  passthru.libraries = callPackages ./libraries.nix versionConfig.libVersion;
+  passthru.libraries = callPackages ./libraries.nix { inherit libSrc libVersion; };
```
```suggestion
  # Common libraries, referenced during runtime, via the wrapper.
  passthru.libraries = callPackages ./libraries.nix { inherit libSrc libVersion; };
```

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -46,29 +49,22 @@ assert ngspiceSupport -> libngspice != null;
-  versionConfig = versions.${baseName};
-
   # oce on aarch64 fails a test
   withOCE = oceSupport && !stdenv.isAarch64;
```
This is not part of the PR, I know. This should be an assertion and not a silent disable.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -46,29 +49,22 @@ assert ngspiceSupport -> libngspice != null;
-
   # oce on aarch64 fails a test
   withOCE = oceSupport && !stdenv.isAarch64;
   withOCC = (withOCCT && !withOCE) || (oceSupport && stdenv.isAarch64);
```
This as well, put the transition to an assertion in a separate PR, please.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -21,11 +21,14 @@
+, i18nVersion
 , oceSupport ? false
 , opencascade
 , withOCCT ? true
```
Using `*Support` variables and `with*` together is misleading, but it's an API that's already exposed. It could be deprecated rather nicely with an assertion as well (`lib.assertions.assertMsg` I think is good).

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -46,29 +49,22 @@ assert ngspiceSupport -> libngspice != null;
 
 with lib;
```
While at it, we'd be better of without this.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,52 +1,165 @@
 stdenv.mkDerivation rec {
 
-  passthru.libraries = callPackages ./libraries.nix versionConfig.libVersion;
+  passthru.libraries = callPackages ./libraries.nix { inherit libSrc libVersion; };
+  passthru.i18n = callPackage ./i18n.nix {
```
OTH, I agree that putting `i18n` outside of libraries makes more sense, since it's a requirement of `base.nix` and not the rest, so... maybe make `passthru.libraries.i18n = throw "this attribute has moved to "kicad-unstable.passthru.i18n"`. Or maybe use `lib.warn`.

matthuszagh replied:
I'm not sure I completely understand this point. Why do we need to pass i18n to libraries at all? Aren't libraries only called through kicad?

doronbehar replied:
It's not a question of whether `libraries.nix` needs `i18n` or not, we know `base.nix` needs it. It's a question of whether we should change the "API" of `kicad-unstable.passsthru`. Someone out there may be relying on it, if they have maybe some other build linking against `kicad-unstable.libraries.i18n`, this PR might tell them to use `passthru.i18n` instead, and I'm just unsure whether this is justified.

doronbehar replied:
We are already doing some backwards incompatible changes. Hence I'd keep things as are :). Sorry for maybe misleading you. I'm thinking as I'm reviewing.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -46,29 +49,22 @@ assert ngspiceSupport -> libngspice != null;
 stdenv.mkDerivation rec {
 
-  i18n = libraries.i18n;
+  i18n = callPackage ./i18n.nix {
+    src = i18nSrc;
+    version = i18nVersion;
+  };
```
This is called twice eventually - not good. I think it fits more if it's in `default.nix` because it's the entry point of whoever that inspects this derivation.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,52 +1,165 @@
 
-  passthru.libraries = callPackages ./libraries.nix versionConfig.libVersion;
+  passthru.libraries = callPackages ./libraries.nix { inherit libSrc libVersion; };
+  passthru.i18n = callPackage ./i18n.nix {
```
Called twice - the 1st / 2nd time here.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,52 +1,165 @@
-    inherit versions stable baseName;
+    inherit stable baseName;
+    inherit kicadSrc kicadVersion;
+    inherit i18nSrc i18nVersion;
```
```suggestion
    inherit (passthru) i18n;
```

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -59,7 +172,7 @@ stdenv.mkDerivation rec {
-      [ python.pkgs.wrapPython ];
+    [ python.pkgs.wrapPython ];
 
   # wrapGAppsHook added the equivalent to ${base}/share
   # though i noticed no difference without it
```
Nitpick:

```suggestion
  # We are emulating wrapGAppsHook, along with other variables to the wrapper
```

evils replied:
i wasn't emulating wrapGAppsHook, since that misbehaves

doronbehar reviewed (commented):
The review is a bit messy sorry, but the expression is getting better and better :). Some suggestions are better of in separate commits.

matthuszagh committed:
kicad: Rename `<feature>Support` arguments to `with<Feature>`

Also: Use assertions instead of silently ignoring arguments that don't cooperate
(occ+oce) / won't compile (aarch64 + oce).

base.nix no longer provides default argument values since these are
provided by default.nix.

matthuszagh committed:
kicad: clarify wrapGAppsHook comment

matthuszagh commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -46,29 +49,22 @@ assert ngspiceSupport -> libngspice != null;
-      repo = "kicad";
-    } // versionConfig.kicadVersion.src
-  );
+  version = kicadVersion;
```
Would you mind clarifying this point? I'm not sure I completely understand...

doronbehar replied:
[I said before](https://github.com/NixOS/nixpkgs/pull/100066#issuecomment-706604693) that this PR triggers a change in the inputs of `kicad-unstable`, and it shouldn't. I said it shouldn't because it's only a change in the way the derivations are evaluated, and not in the derivations themselves. But, this is an example of a _justified_ "change" in one of the dependencies that triggers a rebuild, but isn't much of a big deal (only a change in the value of `version` which affects it's `name`). Hence I take back my words that this PR shouldn't trigger rebuilds - so there's nothing for you to do around this.

matthuszagh committed:
kicad: remove unnecessary with lib; statements

matthuszagh committed:
kicad: add srcs parameter to allow configuring kicad versions

This also exposes the full src and version parameters for each
derivation, allowing them to overrideable by srcs.

matthuszagh commented:
@doronbehar I hope I got everything (I think I did), but sorry in advance if I messed something up (got into a little trouble with all the rebasing :)).

Also let me know about those other two points.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -41,18 +41,17 @@
-assert ngspiceSupport -> libngspice != null;
-
+# oce on aarch64 fails a test
+assert withOCE -> !stdenv.isAarch64;
```
Use `stdenv.lib.assertions.assertMsg`.

https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-asserts

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -28,7 +31,13 @@
 
-assert ngspiceSupport -> libngspice != null;
+assert withNgspice -> libngspice != null;
+assert stdenv.lib.assertMsg (!ngspiceSupport)
+  "nspiceSupport is deprecated. Use withNgspice instead.";
+assert stdenv.lib.assertMsg (!oceSupport)
+  "oceSupport is deprecated. Use withOCE instead.";
+assert stdenv.lib.assertMsg (!scriptingSupport)
+  "scriptingSupport is deprecated. Use withScripting instead.";
```
Excellent.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -44,29 +46,16 @@
-      repo = "kicad";
-    } // versionConfig.kicadVersion.src
-  );
+  inherit i18n;
```
Not needed, as it's used in `postInstall`.

doronbehar reviewed (commented):
Not many more to go. The commit messages look good overall. Only commit https://github.com/NixOS/nixpkgs/commit/239d16c2c20971984f91f083cce5be6f373e5280 should mention the rename of the `*Support` arguments to `with*`.

matthuszagh commented:
I've implemented the changes. Let me know if there's anything else.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -41,18 +41,16 @@
 
-assert ngspiceSupport -> libngspice != null;
-
+assert stdenv.lib.asserts.assertMsg (!(withOCE && stdenv.isAarch64)) "OCE fails a test on Aarch64";
```
```suggestion
assert stdenv.lib.asserts.assertMsg (!(withOCE && stdenv.isAarch64)) "OCE fails a test on Aarch64";
# WHY?
assert stdenv.lib.asserts.assertMsg (withOCCT && !withOCE) "Can't enable OCCT and not OCE";
```

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -41,18 +41,16 @@
-
   libraries = callPackages ./libraries.nix versionConfig.libVersion;
 
+  withOCC = assert (withOCCT && !withOCE) || (withOCE && stdenv.isAarch64); true;
```
The idea was to remove this messy assignment. What is the problem with cross enabling all of these features? I understand the lack of aarch64 support, but not this one. Maybe the expression should be simpler in favor of a failing build?

doronbehar replied:
Also the use of both `withOCC` variable and `withOCCE` is confusing for someone reading this.

matthuszagh replied:
I've taken this statement out and added an assertion to ensure occ and oce are not both enabled. Kicad doesn't actually impose this requirement on us; it simply uses occ if occ and oce are both specified:

```
option( KICAD_USE_OCE
    "Build tools and plugins related to OpenCascade Community Edition (default ON)"
    ON )

option( KICAD_USE_OCC
    "Build tools and plugins related to OpenCascade Technology (overrides KICAD_USE_OCE, default OFF)"
    OFF )
```

So, if you think it's better I can take out the assertion.

doronbehar replied:
No keep it, besides the typo, everything is perfect IMO - this as well.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -25,13 +25,13 @@
 , stable ? true
```
The commit message should be (IMO):

```
kicad: Rename `<feature>Support` arguments to `with<Feature>`

Also: Use assertions instead of silently ignoring arguments that don't cooperate
(occt+oce) / won't compile (aarch64 + oce)
```

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -25,13 +25,13 @@
 , baseName ? "kicad"
 , versions ? { }
-, oceSupport ? false
+, withOCE ? false
```
Slight nitpick: The default values are already inherited from `default.nix` so it shouldn't be needed to specify them here as well, I think.

doronbehar replied:
It would be preferred to put out the defaults in `default.nix` and not list them here as well, as again - `default.nix` is the 1st thing a 3rd party reader reads.

matthuszagh replied:
I removed all defaults in addition to the `with<feature>` ones since they're all provided in default.nix. Let me know if that's too far and I'll add those back in.

doronbehar replied:
No that's perfect.

doronbehar reviewed (commented):
Thanks for your cooperation. Only issues I see are in the assertions commit. It's almost perfect.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -61,7 +70,7 @@ stdenv.mkDerivation rec {
     inherit versions stable baseName;
     inherit wxGTK python wxPython;
-    inherit debug withI18n withOCCT oceSupport ngspiceSupport scriptingSupport;
+    inherit debug withI18n withOCCT withOCE withNgspice withScripting;
```
```suggestion
    inherit debug withI18n withOCC withOCE withNgspice withScripting;
```

doronbehar replied:
Or maybe use `withOCCT` on `base.nix` as well.

matthuszagh replied:
Ah right, forgot about default.nix. Isn't OCCT confusing here though since the kicad cmake parameter is OCC?

matthuszagh replied:
Or, better to keep the interface the same, hence using OCCT?

doronbehar replied:
Yea... It's up to you whether to change this API as well. I would. With a similar deprecation warning as the others.

doronbehar commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -22,37 +22,36 @@
+, withOCE
 , opencascade
-, withOCCT ? true
+, withOCC
```
Same as it used to be.

matthuszagh commented:
I've made the changes, let me know what you think @doronbehar.

doronbehar commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -28,7 +32,15 @@
+assert stdenv.lib.assertMsg (!scriptingSupport)
+  "scriptingSupport is deprecated. Use withScripting instead.";
+assert stdenv.lib.assertMsg (!withOCCT)
+  "withOCCT is deprecated. Use withOCC instead.";
```
On a 2nd thought, now it hits me that we didn't have a deprecation period :laughing: . Personally I'm in favor of doing small incompatible changes like this without warning, but at least we should comply to the definition of the words we are using. So maybe (example):

```
Please use `withOCCT` instead of `withOCC`
```

Or:

```
`withOCCT` was renamed to `withOCC` for the sake of consistency with upstream cmake options
```

And:

```
`scriptingSupport` was renamed to `withScripting` for the sake of consistency with other kicad nix arguments.
```

matthuszagh replied:
You mean take out these deprecation assertions? I'm happy to do that if you think it's better. If we leave them in, how long before I should come back and take them all out?

And don't worry about the many rounds of edits, better to get this right the first time than to have to come back and fix it later :)

doronbehar replied:
I meant only to change the message.

doronbehar replied:
Please leave the assertions as are.

matthuszagh replied:
Got it. Ignore that last force push. I still need to add the release notes.

doronbehar reviewed (changes requested):
Also, to be nice to everyone else, we should add release notes to rl-2103.xml (ideally in seperate commit from this one).

The commit message of this "rename options commit" should mention also the fact that the defaults are not needed in `base.nix`, although it doesn't affect users directly.

1 last round I promise!

doronbehar commented:
I know it ruins the `git commit --fixup=` workflow, but could you please mention in that commit the fact that `base.nix` doesn't need the defaults to be specified in it?

Also, we'll definitely need a commit that mentions these api changes in `rl-2103.xml`.

matthuszagh committed:
add 21.03 release notes for renamed kicad options

matthuszagh commented:
Changes in.

doronbehar reviewed (approved):
Thanks for your contribution.

evils commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,68 +1,192 @@
 
-  versions =  import ./versions.nix;
-  versionConfig = versions.${baseName};
+  libSrcFetch = name: fetchFromGitLab {
```
one of the goals of my rewrite was to simplify default.nix to be as much as possible just the wrapper
code that's related to building the components that get tied together here should be in base.nix or libraries.nix respectively

evils commented on pkgs/applications/science/electronics/kicad/libraries.nix:
```diff
@@ -42,30 +28,4 @@ in
   footprints = mkLib "footprints";
   packages3d = mkLib "packages3d";
-
-  # i18n is a special case, not actually a library
```
i think this description of i18n being optional, separate and only found via a relative path should have been retained

evils reviewed (commented):
sorry for such a late review
i missed it because i confused its mails for #85456
and it being merged just 3 days after being opened did not leave a lot of time to discover that mistake
